graphqshell/test/GraphQL/Schema/IntrospectionTest.hs

```diff
@@ -0,0 +1,65 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE RankNTypes #-}
+
+module GraphQL.Schema.IntrospectionTest where
+
+import Control.Exception.Safe (MonadThrow)
+import Data.Aeson (FromJSON, ToJSON, Value)
+import GraphQL.Client.Types
+import GraphQL.Introspection.Marshalling.Types (IntrospectionResponse)
+import GraphQL.Introspection.Schema
+import GraphQL.Introspection.Schema.Types
+import GraphQL.Schema.Fixtures
+import Relude
+import Test.Tasty ()
+import Test.Tasty.Hspec
+
+-- data TestClientSettings a = TestClientSettings
+--   { -- TODO: map query to response
+--     stubResponse :: GraphQLResponse a
+--   }
+--   deriving (Eq, Show)
+
+newtype TestGraphQLClient r a = TestGraphQLClient
+  { runTestClient :: ReaderT (GraphQLResponse r) IO a
+  }
+  deriving (Functor, Applicative, Monad, MonadIO, MonadThrow, MonadReader (GraphQLResponse r))
+
+withStubbedClient :: (GraphQLResponse IntrospectionResponse) -> (forall m. GraphQLClient m => m a) -> IO a
+withStubbedClient resp action = runReaderT (runTestClient action) resp
+
+instance (Show a) => GraphQLClient (TestGraphQLClient a) where
+  runGraphQLRequest query variables = do
+    r <- ask
+    liftIO $ print r
+    --    pure r
```

Comment:
This however would not type check, because the types can't be deduced. In essence the type we want to return is more general, then the one we create. We have an `(GraphQLResponse IntrospectionResponse)` but we need a `(GraphQLResponse a)`